Week 4 Day 1: Best Practices for Installability (#66)
microsoft/win-student-devs: docs/30DaysOfPWA/platforms-practices/01.md

# 4.1: Installability for PWAs

**Author: Zach Teutsch [@devteutsch](https://twitter.com/devteutsch)**

![Placeholder Banner Only. Replace when final assets ready.](_media/day1.png)

> Welcome to Week 4, Day 1 of the [30 Days of PWA](https://aka.ms/learn-pwa/30Days-blog) series! Now that we've seen how to start building PWAs and what basic and advanced capabilites they bring, it's time to dive into what a _good_ PWA really looks like it, and how we can start to build quality applications ourselves. Week 4 will be all about some of the best practices and technologies you can use to really polish up your PWA.

Reviewer: Maybe it's just me but "start building PWAs" feels like we're at the beginning of the series, not the end.
Not sure how to rephrase it, but I'd want this intro to say something like we've gone through all the steps, from learning basic blocks, to coding, to debugging and testing, and finally packaging for stores. So now it's time to wrap things up with best practices. 
Does that make sense?

![Twitter PWA open on Windows](_media/twitter-pwa.png)

Some of the things that I like about Twitter's PWA:

**Smooth, app-like user experience**

Twitter's PWA doesn't show a lot of the behaviors that many typical websites would:
* It uses the `standalone` [display mode](../advanced-capabilities/04.md) to present the the window as an app. 
* The view never refreshes, the shell UI is always visible and only the main content area changes with user interaction. 
* There's very little latency on navigation. Fetched content (like the tweets themselves) may take a second to load, but the UI interactivity itself is nearly instant.

We'll learn more about best practices for PWA UI/UX on [Day 7](07.md) of Platforms and practices week.

**Consistent offline behavior**

Even though Twitter relies pretty heavily on connectivity, it still maintains as much of the app interaction as possible when connection is lost (just like any native app would).

When connection is lost, the app still loads its view and the UI is still navigable. If there are any tweets stored locally, those will be displayed. If there aren't, the user will get a notice that doesn't interfere with the rest of the UI:

> ![](_media/twitter-offline.png#test)

Reviewer: Missing alt text here.

Author: ```suggestion
![Twitter offline warning](_media/twitter-offline.png#test)
```

We already learned a little about offline behavior on [Day 5](../core-concepts/05.md) of Week 1 and more is coming in the [reliability article](04.md) later this week.

**OS integration and plenty of web capabilities**

Twitter also makes great use of available web capabilities to integrate with your device. Just to list some:

* Push notifications
* Sharing from Twitter
* Sharing to Twitter
* Badging
* Shortcuts

> 

Reviewer 2: Maybe add a screenshot of the Twitter shortcuts or something? Just something to drive this point home some more.

Author: I like this idea.

We already covered how to use all these capabilities and more in [Week 2: Advanced Capabilities.](../advanced-capabilities) Go check them out!

**Published to the Microsoft Store**

> Twitter takes advantage of one the best parts about PWAs: discoverability on the web _and_ in app stores.

Reviewer 2: Maybe change this to `discoverability and install-ability` just to point out that PWAs can be installed from the browser too.

Author: ```suggestion
Twitter takes advantage of one the best parts about PWAs: discoverability and installability on the web _and_ in app stores.
```

![Twitter in the Microsoft Store](_media/twitter-in-store.png)

> The same PWA the can be installed from the browser can be distributed in app stores. You can learn more about packaging for the stores in our [article on PWABuilder](../dev-tools/02.md) and more info about best practices for distribution to different app stores is coming [later this week.](03.md)

Reviewer: I think you probably want to link to day 7 of week 2, not day 2, right?

Author: ```suggestion
The same PWA the can be installed from the browser can be distributed in app stores. You can learn more about packaging for the stores in our [article on PWABuilder](../dev-tools/07.md) and more info about best practices for distribution to different app stores is coming [later this week.](03.md)
```

> ## Summary

Reviewer: Before the summary, what do you think about showing other great PWAs too. I don't think we should go in as much details as what you've done for Twitter (which is great, btw), but adding a couple of other telling examples might be good for people to get an even better understanding of what a great PWA is.
https://www.pwastats.com/ might provide a good list of inspiration for this. I can also think of Photopea, and Adobe Creative Cloud Express.

Reviewer 2: Clipchamp is another good one

Author: @captainbrosset @jgw96  My original plan was to show multiple PWAs (including clipchamp), however, I found that driving home a single PWA is helpful to showcase a top-to-bottom experience, as opposed to cherry picking features from different apps. 

It works well for pushing the idea that adding a web capability here and there doesn't necessarily make a good PWA.

# 4.1: Installability for PWAs

**Author: Zach Teutsch [@devteutsch](https://twitter.com/devteutsch)**

> Welcome to Week 4, Day 1 of the [30 Days of PWA](https://aka.ms/learn-pwa/30Days-blog) series! Now that we've seen how to start building PWAs and what basic and advanced capabilites they bring, it's time to dive into what a _good_ PWA really looks like it, and how we can turn our PWAs into quality application experiences. Week 4 is all about some of the best practices and technologies you can use to really polish up your PWA.

Reviewer 3: ```suggestion
Welcome to Week 4, Day 1 of the [30 Days of PWA](https://aka.ms/learn-pwa/30Days-blog) series! Now that we've seen how to [start building PWAs](../core-concepts/) and what basic and [advanced capabilities](../advanced-capabilities/) they bring, it's time to dive into what a _good_ PWA really looks like, and how we can turn our PWAs into quality application experiences. Week 4 is all about some of the best practices and technologies you can use to really polish up your PWA.
```

## Summary

At the end of the day, best practices for PWA installability are really just about best practices for PWAs in general. When you prompt a user to install your app to their machine, that app needs to live up to the expectations of applications on that platform. Luckily enough, we're going to spend all week building on what we've already learned about PWAs and seeing some best practices for making solid PWA experiences. If you're looking to get a head start, check out the [PWA Starter](https://aka.ms/learn-pwa/30days-4.1/github.com/pwa-builder/pwa-starter) for a PWA template that's built with best practices in mind.

### **Exercise**

Think about some of your favorite web experiences that you use or have built yourself. 

Would any of these experiences make good PWAs? Would any of them not? 

What could these experiences gain by taking advantage of some of the capabilites that PWAs bring to the table?

### **Resources**

* **[Microsoft PWA Overview](https://aka.ms/learn-pwa/30days-4.1/docs.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium)**

> * **[PWA Advanced Capabilities Week](../advanced-capabilities)**

Reviewer 3: ```suggestion
* **[PWA Advanced Capabilities Week](../advanced-capabilities/)**
```

For some reason the generated site won't redirect without the `/`.
